datastore/longhorn.go

```diff
@@ -2961,3 +3066,153 @@ func (s *DataStore) RemoveFinalizerForBackup(backup *longhorn.Backup) error {
+	if err != nil {
+		return nil, err
+	}
+
+	for _, itemRO := range list {
+		// Cannot use cached object from lister
+		itemMap[itemRO.Name] = itemRO.DeepCopy()
+	}
+	return itemMap, nil
+}
+
+func (s *DataStore) getRecurringJobRO(name string) (*longhorn.RecurringJob, error) {
+	return s.rjLister.RecurringJobs(s.namespace).Get(name)
+}
+
+// GetRecurringJob gets the RecurringJob for the given name and namespace.
+// Returns a mutable RecurringJob object
+func (s *DataStore) GetRecurringJob(name string) (*longhorn.RecurringJob, error) {
+	result, err := s.getRecurringJob(name)
+	if err != nil {
+		return nil, err
+	}
+	return result.DeepCopy(), nil
+}
+
+func (s *DataStore) getRecurringJob(name string) (*longhorn.RecurringJob, error) {
+	return s.rjLister.RecurringJobs(s.namespace).Get(name)
+}
+
+// UpdateRecurringJob updates Longhorn RecurringJob and verifies update
+func (s *DataStore) UpdateRecurringJob(recurringJob *longhorn.RecurringJob) (*longhorn.RecurringJob, error) {
+	obj, err := s.lhClient.LonghornV1beta1().RecurringJobs(s.namespace).Update(recurringJob)
+	if err != nil {
+		return nil, err
+	}
+	verifyUpdate(recurringJob.Name, obj, func(name string) (runtime.Object, error) {
+		return s.getRecurringJob(name)
+	})
+	return obj, nil
+}
+
+// UpdateRecurringJobStatus updates Longhorn RecurringJob resource status and
+// verifies update
+func (s *DataStore) UpdateRecurringJobStatus(recurringJob *longhorn.RecurringJob) (*longhorn.RecurringJob, error) {
+	obj, err := s.lhClient.LonghornV1beta1().RecurringJobs(s.namespace).UpdateStatus(recurringJob)
+	if err != nil {
+		return nil, err
+	}
+	verifyUpdate(recurringJob.Name, obj, func(name string) (runtime.Object, error) {
+		return s.getRecurringJobRO(name)
+	})
+	return obj, nil
+}
+
+// DeleteRecurringJob deletes the RecurringJob.
+// The dependents will be deleted in the foreground
+func (s *DataStore) DeleteRecurringJob(name string) error {
+	propagation := metav1.DeletePropagationForeground
+	return s.lhClient.LonghornV1beta1().RecurringJobs(s.namespace).Delete(
+		name, &metav1.DeleteOptions{PropagationPolicy: &propagation},
+	)
+}
+
+// RemoveFinalizerForRecurringJob will result in deletion if DeletionTimestamp was set
+func (s *DataStore) RemoveFinalizerForRecurringJob(obj *longhorn.RecurringJob) error {
+	if !util.FinalizerExists(longhornFinalizerKey, obj) {
+		// finalizer already removed
+		return nil
+	}
+	if err := util.RemoveFinalizer(longhornFinalizerKey, obj); err != nil {
+		return err
+	}
+	_, err := s.lhClient.LonghornV1beta1().RecurringJobs(s.namespace).Update(obj)
+	if err != nil {
+		// workaround `StorageError: invalid object, Code: 4` due to empty object
+		if obj.DeletionTimestamp != nil {
+			return nil
+		}
+		return errors.Wrap(err, "unable to remove finalizer for recurring job")
+	}
+	return nil
+}
+
+func ValidateRecurringJob(job types.RecurringJobSpec) error {
+	if job.Cron == "" || job.Task == "" || job.Name == "" || job.Retain == 0 || job.Concurrency == 0 {
```

Comment:
> 1. if `job.Concurrency == 0`, it's better to consider it as the default value. Since the old recurring jobs in volumeSpec or StorageClass don't contain this field. Or users may not care about this field at all.

Fixed.

> 2. Need to check `Task`

Fixed.